[scroll-animations] implement parsing support for `view-timeline` and its longhand properties

graouts opened the pull request:
#### 7fc2fe510e84f36506e903d14dbbddeae2dc3826
<pre>
[scroll-animations] implement parsing support for `view-timeline` and its longhand properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=265689">https://bugs.webkit.org/show_bug.cgi?id=265689</a>

Reviewed by Tim Nguyen.

Add parsing support for the `view-timeline` shorthand and its three coordinating list base
property longhands `view-timeline-name`, `view-timeline-axis` and `view-timeline-inset` as
specified in <a href="https://drafts.csswg.org/scroll-animations-1/#view-timelines-named.">https://drafts.csswg.org/scroll-animations-1/#view-timelines-named.</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-axis-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-axis-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-inset-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-inset-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-name-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-name-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-shorthand-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::create):
(WebCore::ViewTimeline::ViewTimeline):
* Source/WebCore/animation/ViewTimeline.h:
(WebCore::ViewTimelineInsets::operator== const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForSingleViewTimelineInset):
(WebCore::valueForViewTimelineInset):
(WebCore::viewTimelineShorthandValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::CSSPropertyParser::consumeScrollTimelineShorthand):
(WebCore::CSSPropertyParser::consumeViewTimelineShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTimelineInsetListItem):
(WebCore::CSSPropertyParserHelpers::consumeViewTimelineInset):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::adjustViewTimelines):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialViewTimelineAxes):
(WebCore::RenderStyle::initialViewTimelineInsets):
(WebCore::RenderStyle::initialViewTimelineNames):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::viewTimelines const):
(WebCore::RenderStyle::viewTimelineAxes const):
(WebCore::RenderStyle::viewTimelineInsets const):
(WebCore::RenderStyle::viewTimelineNames const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setViewTimelineAxes):
(WebCore::RenderStyle::setViewTimelineInsets):
(WebCore::RenderStyle::setViewTimelineNames):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertViewTimelineInset):

Canonical link: <a href="https://commits.webkit.org/271446@main">https://commits.webkit.org/271446@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa43b68d926cafef48df3afcc96688bbf135002b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28438 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7083 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29824 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30965 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25896 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28935 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4452 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28707 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5843 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24467 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5112 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25468 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31651 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25909 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31511 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5183 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29281 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6783 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5640 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3666 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2661,6 +2665,54 @@ bool CSSPropertyParser::consumeScrollTimelineShorthand(bool important)
+        else
+            return false;
+
+        // Both a view-timeline-axis and a view-timeline-inset are optional.
```
This comment is a bit misleading, since one of two is required (according to the spec and what you implemented)

graouts replied:
From my reading of the spec for [view-timeline](https://drafts.csswg.org/scroll-animations-1/#propdef-view-timeline) the axis and insets are optional: `[ <'view-timeline-name'> [ <'view-timeline-axis'> || <'view-timeline-inset'> ]? ]#`. That certainly is what I implemented as well. What makes you think otherwise?

nt1m replied:
Oh nevermind, I misread your code and the spec too. I missed the `?` from the spec.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2661,6 +2665,54 @@ bool CSSPropertyParser::consumeScrollTimelineShorthand(bool important)
+            if (insets)
+                insetsList.append(insets.releaseNonNull());
+            else
+                insetsList.append(defaultInsets());
```
```suggestion
            axesList.append(axis ? axis.releaseNonNull() : defaultAxis());
            insetsList.append(insets ? insets.releaseNonNull() : defaultInsets());
```
This ternary would be easier to read IMO

graouts replied:
I agree, and in fact is what I wrote originally but I couldn't work out how to fix the resulting build error:

```
/Users/antoine/Code/safari/OpenSource/Source/WebCore/css/parser/CSSPropertyParser.cpp:2693:34: error: conditional expression is ambiguous; 'Ref<WebCore::CSSValue>' can be converted to 'Ref<CSSPrimitiveValue>' and vice versa
 2693 |             axesList.append(axis ? axis.releaseNonNull() : defaultAxis());
```

But taking a closer took now, if I force `defaultAxis` and `defaultInsets` to be typed as `Ref<CSSValue>` then all is well. So this change is also required:

```diff
-    auto defaultAxis = []() { return CSSPrimitiveValue::create(CSSValueBlock); };
-    auto defaultInsets = []() { return CSSPrimitiveValue::create(CSSValueAuto); };
+    auto defaultAxis = []() -> Ref<CSSValue> { return CSSPrimitiveValue::create(CSSValueBlock); };
+    auto defaultInsets = []() -> Ref<CSSValue> { return CSSPrimitiveValue::create(CSSValueAuto); };
```

I'll implement that.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2096,5 +2098,20 @@ inline Vector<ScrollAxis> BuilderConverter::convertScrollTimelineAxis(BuilderSta
+inline Vector<ViewTimelineInsets> BuilderConverter::convertViewTimelineInset(BuilderState& state, const CSSValue& value)
+{
+    // While parsing, consumeViewTimelineInset() and consumeViewTimelineShorthand() yield a CSSValueList exclusively.
+    ASSERT(is<CSSValueList>(value));
```
CSS Typed OM can give random values, we've had fuzz bugs about this. We should probably just early return with the default value if we don't have the default value.

graouts replied:
I don't follow, how does the CSS Typed OM relate to this code? My understanding is that the `CSSValue` here stems only from the result of parsing the `view-timeline-inset` longhand or `view-timeline` shorthand. What am I missing?

nt1m replied:
```
let cssStyleValue = CSSMathValue.parseAll('columns', 'auto')[0];
document.body.attributeStyleMap.append('view-timeline-inset', cssStyleValue);
```

This JS goes to the style builder converter directly without going through the parsing code.

graouts replied:
I see! So, what do you suggest for this specific function?

graouts replied:
I suppose we simply can't `ASSERT()` the way I am right now, but we need to always check we have the expected input. I'll fix this!

nt1m replied:
What you did in the last revision looks good.

graouts committed:
[scroll-animations] implement parsing support for `view-timeline` and its longhand properties
https://bugs.webkit.org/show_bug.cgi?id=265689

Reviewed by Tim Nguyen.

Add parsing support for the `view-timeline` shorthand and its three coordinating list base
property longhands `view-timeline-name`, `view-timeline-axis` and `view-timeline-inset` as
specified in https://drafts.csswg.org/scroll-animations-1/#view-timelines-named.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-axis-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-axis-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-inset-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-inset-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-name-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-name-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-shorthand-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::create):
(WebCore::ViewTimeline::ViewTimeline):
* Source/WebCore/animation/ViewTimeline.h:
(WebCore::ViewTimelineInsets::operator== const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForSingleViewTimelineInset):
(WebCore::valueForViewTimelineInset):
(WebCore::viewTimelineShorthandValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::CSSPropertyParser::consumeScrollTimelineShorthand):
(WebCore::CSSPropertyParser::consumeViewTimelineShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTimelineInsetListItem):
(WebCore::CSSPropertyParserHelpers::consumeViewTimelineInset):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::adjustViewTimelines):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialViewTimelineAxes):
(WebCore::RenderStyle::initialViewTimelineInsets):
(WebCore::RenderStyle::initialViewTimelineNames):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::viewTimelines const):
(WebCore::RenderStyle::viewTimelineAxes const):
(WebCore::RenderStyle::viewTimelineInsets const):
(WebCore::RenderStyle::viewTimelineNames const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setViewTimelineAxes):
(WebCore::RenderStyle::setViewTimelineInsets):
(WebCore::RenderStyle::setViewTimelineNames):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertViewTimelineInset):

Canonical link: https://commits.webkit.org/271446@main
